src/adapters/mock/adapter.ts

```diff
@@ -32,16 +29,88 @@ export async function mockFetch({
   headers = {},
   body,
 }: NormalizedRequest): Promise<NormalizedResponse> {
-  const resp = await fetch(url, {method, headers: flatHeaders(headers), body});
-  const respBody = await resp.text();
-  return {
-    statusCode: resp.status,
-    statusText: resp.statusText,
-    body: respBody,
-    headers: canonicalizeHeaders(Object.fromEntries(resp.headers.entries())),
-  };
+  canonicalizeHeaders(headers);
+  const matchingRequest = mockAdapter.findRequest({
+    url,
+    method,
+    headers,
+    body,
+  });
+  if (typeof matchingRequest === 'undefined') {
+    mockAdapter.requestList.push({url, method, headers, body, attempts: 1});
+  } else {
+    matchingRequest.attempts++;
+  }
+
+  const next = mockAdapter.responseList.shift()!;
+  if (next instanceof Error) {
+    throw next;
+  }
+  return next;
+}
+
+type RequestListEntry = NormalizedRequest & {attempts: number};
+type ResponseListEntry = NormalizedResponse | Error;
+
+interface MockedAdapter {
+  requestList: RequestListEntry[];
+  responseList: ResponseListEntry[];
+  getOldestRequest: () => NormalizedRequest;
+  getMostRecentRequest: () => NormalizedRequest;
+  queueResponse: (response: NormalizedResponse) => void;
+  queueError: (error: Error) => void;
+  findRequest: (request: NormalizedRequest) => RequestListEntry | undefined;
+  reset: () => void;
 }
 
+export const mockAdapter: MockedAdapter = {
+  requestList: [],
+  responseList: [],
+
+  getOldestRequest(): NormalizedRequest {
+    if (this.requestList.length === 0) {
+      throw new Error('No requests have been made');
+    }
+    return this.requestList.shift()!;
+  },
+
+  getMostRecentRequest(): NormalizedRequest {
+    if (this.requestList.length === 0) {
+      throw new Error('No requests have been made');
+    }
+    return this.requestList.pop()!;
+  },
+
+  queueResponse(response: NormalizedResponse): void {
+    this.responseList.push(response);
+  },
+
+  queueError(error: Error): void {
+    this.responseList.push(error);
+  },
+
+  findRequest(request: NormalizedRequest): RequestListEntry | undefined {
+    for (const matchingRequest of this.requestList) {
+      if (
+        matchingRequest.url === request.url &&
+        matchingRequest.method === request.method
+      ) {
+        if (request.body === undefined || request.body === null) {
+          return matchingRequest;
+        } else if (matchingRequest.body === request.body) {
+          return matchingRequest;
+        }
```

Comment:
This would be a test setup error, but what if we try to find a request that should have a body but doesn't? Could we check that the body matches regardless of it being null?

```suggestion
        if (!request.body && !matchingRequest.body) {
          return matchingRequest;
        } else if (matchingRequest.body === request.body) {
          return matchingRequest;
        }
```